pkg/blockdevice/v1alpha2/blockdevice.go

```diff
@@ -0,0 +1,178 @@
+//	}
+//}
+
+// IsType returns true if the block device is of type same as passed argument
+//func IsType(bdType string) Predicate {
+//	return func(d *BlockDevice) bool {
+//		return d.Object.GetLabels()[string(apisv1alpha1.NdmBlockDeviceTypeCPK)] == bdType
+//	}
+//}
+
+// IsActive returns true if the block device is active.
+func IsActive() Predicate {
+	return func(bd *BlockDevice) bool {
+		return bd.Object.Status.State == string(BlockDeviceStateActive)
+	}
+}
+
+// IsUsable returns true if this block device can be used for pool provisioning.
+// The argument usedBlockDevice is a map containing key as block device cr name and value as integer.
+// If the value of map is greater than 0 , then this corresponding block device is not usable.
+func IsUsable(usedBlockDevices map[string]int) Predicate {
+	return func(bd *BlockDevice) bool {
+		return usedBlockDevices[bd.Object.Name] == 0
+	}
+}
+
+// IsUsableNode returns true if block device of this node can be used for pool provisioning.
+// The argument usedNodes is a map containing key as node name and value as bool.
+// If the value of map is greater than false, then this corresponding node is not usable.
+func IsUsableNode(usedNodes map[string]bool) Predicate {
+	return func(bd *BlockDevice) bool {
+		return !usedNodes[bd.GetNodeName()]
+	}
+}
+
+// IsBelongToNode returns true if the block device belongs to the provided node.
+func IsBelongToNode(nodeName string) Predicate {
+	return func(bd *BlockDevice) bool {
+		return bd.GetNodeName() == nodeName
+	}
+}
+
+// IsValidPoolTopology returns true if the topology is valid.
+func IsValidPoolTopology(poolType string, diskCount int) bool {
+	return DefaultBlockDeviceCount[poolType]%diskCount == 0
+}
+
+// GetNodeName returns the node name to which the block device is attached
+func (bd *BlockDevice) GetNodeName() string {
+	return bd.Object.GetLabels()[string(apisv1alpha1.HostNameCPK)]
+}
+
+// Filter will filter the block device instances if all the predicates succeed
+// against that block device.
+func (l *BlockDeviceList) Filter(p ...Predicate) *BlockDeviceList {
+	var plist predicateList
+	plist = append(plist, p...)
+	if len(plist) == 0 {
+		return l
+	}
+
+	filtered := NewListBuilder().List()
+	for _, bdAPI := range l.ObjectList.Items {
+		bdAPI := bdAPI // pin it
+		BlockDevice := BuilderForAPIObject(&bdAPI).BlockDevice
+		if plist.all(BlockDevice) {
+			filtered.ObjectList.Items = append(filtered.ObjectList.Items, *BlockDevice.Object)
+		}
+	}
+	return filtered
+}
+
+// GetDeviceID returns the device link of the block device. If device link is not found it returns device path.
+// For a cstor pool creation -- this link or path is used. For convenience, we call it as device ID.
+// Hence, device ID can either be a  device link or device path depending on
+// what was available in block device cr.
+func (bd *BlockDevice) GetDeviceID() string {
```

Comment:
Can we rename this to `GetLinkOrPath`
caller code will look like below:
`bd.GetLinkOrPath()`

Reply:
Property in BlockDevice CR is `DeviceID` so it is written as `GetDeviceID`